packages/tailwindcss/src/compat/plugin-api.ts

```diff
@@ -115,6 +115,16 @@ export function buildPluginApi({
         )
       }
 
+      // Ignore variants emitting v3 `:merge(…)` rules. In v4, the `group-*` and `peer-*` variants
+      // compound automatically.
+      if (typeof variant === 'string') {
+        if (variant.includes(':merge(')) return
+      } else if (Array.isArray(variant)) {
+        if (variant.some((v) => v.includes(':merge('))) return
+      } else if (typeof variant === 'object') {
+        if (Object.keys(variant).some((v) => v.includes(':merge('))) return
+      }
```

Comment:
1. We should detect this in matchVariant as well.
2. Should the object check happen recursively or no? I don't remember what v3 did there

Reply:
1. Ah yep forgot about `matchVariant`.
2. Hmm yeah probably you're right.

Reply:
@thecrypticace The `matchVariant(…)` implementation is a bit annoying since we have to call the matcher function to sample the output. I tried throwing the variant out later but once we register it with the design system, the `peer-*` and `group-*` compounding doesn't work anymore. 

The only other idea I have is to check for wether the variant starts with `peer-` or `group-`. What do you think?